File: typescript/src/sync/DeviceLink.ts
/**
 * A signed device-membership record. A user links a new device by having their
 * long-term Ed25519 identity key sign the new device's own public key; every
 * other device verifies that signature to admit the newcomer into the "self"
 * device set — no central directory, no server. Because Ed25519 signatures are
 * deterministic, the serialized record is byte-identical across SDKs (verified
 * against fixtures/sync/vectors.json).
 *
 * Signed body: version(u8=1) · device_id(u16 len + utf8) · device_public_key(32)
 * · issued_at_ms(i64 LE). Serialized form is the signed body followed by the
 * 64-byte Ed25519 signature.
 *
 * SPDX-License-Identifier: MIT
 */

import { Ed25519Service } from "../security/Ed25519Service.js";

/** A signed device-membership record. */
export interface DeviceLink {
  /** The linked device's identifier. */
  deviceId: string;
  /** The device's own 32-byte Ed25519 public key. */
  devicePublicKey: Uint8Array;
  /** When the link was issued (Unix ms; i64). */
  issuedAtMs: bigint;
  /** 64-byte Ed25519 signature by the user's identity key over the signed body. */
  signature: Uint8Array;
}

/** Wire format version; readers reject any other value. */
export const DEVICE_LINK_VERSION = 0x01;

/**
 * The canonical signed body (everything but the signature). Signer and verifier
 * operate over exactly these bytes.
 */
export function signedBody(
  deviceId: string,
  devicePublicKey: Uint8Array,
  issuedAtMs: bigint,
): Uint8Array {
  if (devicePublicKey.length !== 32) {
    throw new Error("Device public key must be 32 bytes.");
  }
  const id = new TextEncoder().encode(deviceId);
  if (id.length > 0xffff) throw new Error("DeviceId is too long.");

  const body = new Uint8Array(1 + 2 + id.length + 32 + 8);
  const dv = new DataView(body.buffer);
  let o = 0;
  body[o++] = DEVICE_LINK_VERSION;
  dv.setUint16(o, id.length, true); o += 2;
  body.set(id, o); o += id.length;
  body.set(devicePublicKey, o); o += 32;
  dv.setBigInt64(o, BigInt.asIntN(64, issuedAtMs), true);
  return body;
}

/** Creates a device-link signed by the user's 32-byte Ed25519 identity seed. */
export function createDeviceLink(
  deviceId: string,
  devicePublicKey: Uint8Array,
  issuedAtMs: bigint,
  identitySeed: Uint8Array,
): DeviceLink {
  const body = signedBody(deviceId, devicePublicKey, issuedAtMs);
  const signature = Ed25519Service.sign(identitySeed, body);
  return { deviceId, devicePublicKey, issuedAtMs, signature };
}

/**
 * True if `link` was signed by the identity behind `identityPublicKey` — i.e.
 * this device belongs to that user.
 */
export function verifyDeviceLink(link: DeviceLink, identityPublicKey: Uint8Array): boolean {
  if (link.signature?.length !== 64) return false;
  if (link.devicePublicKey?.length !== 32) return false;
  const body = signedBody(link.deviceId, link.devicePublicKey, link.issuedAtMs);
  return Ed25519Service.verify(identityPublicKey, body, link.signature);
}

/** Serializes a link as its signed body followed by the 64-byte signature. */
export function serializeDeviceLink(link: DeviceLink): Uint8Array {
  if (link.signature?.length !== 64) throw new Error("Signature must be 64 bytes.");
  const body = signedBody(link.deviceId, link.devicePublicKey, link.issuedAtMs);
  const buf = new Uint8Array(body.length + 64);
  buf.set(body, 0);
  buf.set(link.signature, body.length);
  return buf;
}

/** Parses a serialized link, validating framing. */
export function deserializeDeviceLink(data: Uint8Array): DeviceLink {
  if (data.length < 1 + 2 + 32 + 8 + 64) throw new Error("DeviceLink is too short.");
  const dv = new DataView(data.buffer, data.byteOffset, data.length);
  let o = 0;
  if (data[o++] !== DEVICE_LINK_VERSION) throw new Error("Unsupported DeviceLink format version.");

  const idLen = dv.getUint16(o, true); o += 2;
  if (o + idLen + 32 + 8 + 64 > data.length) throw new Error("DeviceLink is truncated.");
  const deviceId = new TextDecoder().decode(data.subarray(o, o + idLen)); o += idLen;
  const devicePublicKey = data.slice(o, o + 32); o += 32;
  const issuedAtMs = dv.getBigInt64(o, true); o += 8;
  const signature = data.slice(o, o + 64);

  return { deviceId, devicePublicKey, issuedAtMs, signature };
}

File: rust/src/sync/device_link.rs
// SPDX-License-Identifier: MIT

//! A signed device-membership record. A user links a new device by having their
//! long-term Ed25519 identity key sign the new device's own public key; every
//! other device verifies that signature to admit the newcomer into the "self"
//! device set — no central directory, no server. Because Ed25519 signatures are
//! deterministic, the serialized record is byte-identical across SDKs (verified
//! against `fixtures/sync/vectors.json`).
//!
//! Signed body layout (v1): `version(u8=1) · device_id(u16 len + utf8)
//! · device_public_key(32) · issued_at_ms(i64 LE)`. The serialized link is the
//! signed body followed by the 64-byte Ed25519 signature.

use std::fmt;

use crate::security::ed25519::Ed25519SigningService;

/// Wire format version; readers reject any other value.
pub const FORMAT_VERSION: u8 = 0x01;

/// A signed device-membership record.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DeviceLink {
    /// The linked device's identifier.
    pub device_id: String,
    /// The device's own 32-byte Ed25519 public key.
    pub device_public_key: [u8; 32],
    /// When the link was issued (Unix ms).
    pub issued_at_ms: i64,
    /// 64-byte Ed25519 signature by the user's identity key over the signed body.
    pub signature: [u8; 64],
}

/// A framing/validation error from [`create`], [`serialize`] or [`deserialize`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DeviceLinkError {
    /// `device_id` was longer than `u16::MAX` bytes.
    DeviceIdTooLong,
    /// Signing failed (bad identity private key length).
    SignFailed,
    /// The buffer was shorter than the fixed minimum framing.
    TooShort,
    /// The version byte was not [`FORMAT_VERSION`].
    UnsupportedVersion(u8),
    /// A length prefix ran past the end of the buffer.
    Truncated,
    /// A UTF-8 field was not valid UTF-8.
    InvalidUtf8,
}

impl fmt::Display for DeviceLinkError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            DeviceLinkError::DeviceIdTooLong => write!(f, "DeviceId is too long (> u16::MAX bytes)"),
            DeviceLinkError::SignFailed => write!(f, "failed to sign DeviceLink body"),
            DeviceLinkError::TooShort => write!(f, "DeviceLink is too short"),
            DeviceLinkError::UnsupportedVersion(v) => {
                write!(f, "unsupported DeviceLink format version: {v}")
            }
            DeviceLinkError::Truncated => write!(f, "DeviceLink is truncated"),
            DeviceLinkError::InvalidUtf8 => write!(f, "DeviceLink device_id is not valid UTF-8"),
        }
    }
}

impl std::error::Error for DeviceLinkError {}

// Fixed minimum: version(1) + device_id len(2) + device_public_key(32)
// + issued_at_ms(8) + signature(64).
const MIN_LEN: usize = 1 + 2 + 32 + 8 + 64;

/// The canonical signed body (everything but the signature): `version ·
/// device_id(u16 len + utf8) · device_public_key(32) · issued_at_ms(i64 LE)`.
/// Signer and verifier operate over exactly these bytes.
pub fn signed_body(
    device_id: &str,
    device_public_key: &[u8; 32],
    issued_at_ms: i64,
) -> Result<Vec<u8>, DeviceLinkError> {
    let id = device_id.as_bytes();
    if id.len() > u16::MAX as usize {
        return Err(DeviceLinkError::DeviceIdTooLong);
    }

    let mut body = Vec::with_capacity(1 + 2 + id.len() + 32 + 8);
    body.push(FORMAT_VERSION);
    body.extend_from_slice(&(id.len() as u16).to_le_bytes());
    body.extend_from_slice(id);
    body.extend_from_slice(device_public_key);
    body.extend_from_slice(&issued_at_ms.to_le_bytes());
    Ok(body)
}

/// Creates a device-link signed by the user's 32-byte Ed25519 identity private key.
pub fn create(
    device_id: &str,
    device_public_key: [u8; 32],
    issued_at_ms: i64,
    identity_private_key: &[u8],
) -> Result<DeviceLink, DeviceLinkError> {
    let body = signed_body(device_id, &device_public_key, issued_at_ms)?;
    let sig_vec = Ed25519SigningService::sign(identity_private_key, &body)
        .map_err(|_| DeviceLinkError::SignFailed)?;
    let signature: [u8; 64] = sig_vec
        .as_slice()
        .try_into()
        .map_err(|_| DeviceLinkError::SignFailed)?;
    Ok(DeviceLink {
        device_id: device_id.to_string(),
        device_public_key,
        issued_at_ms,
        signature,
    })
}

/// True if `link` was signed by the identity behind `identity_public_key` —
/// i.e. this device belongs to that user.
pub fn verify(link: &DeviceLink, identity_public_key: &[u8]) -> bool {
    let Ok(body) = signed_body(&link.device_id, &link.device_public_key, link.issued_at_ms) else {
        return false;
    };
    Ed25519SigningService::verify(identity_public_key, &body, &link.signature)
}

/// Serializes a link as its signed body followed by the 64-byte signature.
pub fn serialize(link: &DeviceLink) -> Result<Vec<u8>, DeviceLinkError> {
    let mut body = signed_body(&link.device_id, &link.device_public_key, link.issued_at_ms)?;
    body.extend_from_slice(&link.signature);
    Ok(body)
}

/// Parses a serialized link, validating framing.
pub fn deserialize(data: &[u8]) -> Result<DeviceLink, DeviceLinkError> {
    if data.len() < MIN_LEN {
        return Err(DeviceLinkError::TooShort);
    }
    let mut o = 0usize;

    if data[o] != FORMAT_VERSION {
        return Err(DeviceLinkError::UnsupportedVersion(data[o]));
    }
    o += 1;

    let id_len = u16::from_le_bytes(data[o..o + 2].try_into().unwrap()) as usize;
    o += 2;
    // Everything after the id: public key(32) + issued_at(8) + signature(64).
    if o + id_len + 32 + 8 + 64 > data.len() {
        return Err(DeviceLinkError::Truncated);
    }
    let device_id =
        String::from_utf8(data[o..o + id_len].to_vec()).map_err(|_| DeviceLinkError::InvalidUtf8)?;
    o += id_len;

    let mut device_public_key = [0u8; 32];
    device_public_key.copy_from_slice(&data[o..o + 32]);
    o += 32;

    let issued_at_ms = i64::from_le_bytes(data[o..o + 8].try_into().unwrap());
    o += 8;

    let mut signature = [0u8; 64];
    signature.copy_from_slice(&data[o..o + 64]);

    Ok(DeviceLink {
        device_id,
        device_public_key,
        issued_at_ms,
        signature,
    })
}
